Sources/Datadog/URLSessionAutoInstrumentation/URLSessionSwizzler.swift

```diff
@@ -78,7 +78,7 @@ internal class URLSessionSwizzler {
                         var taskReference: URLSessionDataTask?
                         let newCompletionHandler: CompletionHandler = { data, response, error in
                             if let task = taskReference { // sanity check, should always succeed
-                                interceptor.taskCompleted(urlSession: session, task: task, error: error)
```

Comment:
we have this in line 73:
```swift
guard let interceptor = (session.delegate as? DDURLSessionDelegate)?.interceptor else {
  return previousImplementation(session, Self.selector, urlRequest, completionHandler)
}
```
that's how we check _if the session is autotraced or not_.
without this check, we start traces/RUM resources and never finish them.

the problem with `AF` was that it sets its own `session.delegate` privately and doesn't let us provide our own `URLSessionDelegate` in any way. IOW `session.delegate` is **never** a `DDURLSessionDelegate` instance.
this leads us **not** to autotrace the session.

even if we let our users implement their own `InterceptorType`, we won't be calling their `interceptor`s as their `session.delegate` is `Alamofire.SessionDelegate` type.

i hope i'm missing a point 🤞

Reply:
I know this logic, but it has nothing to do with this PR.

As you say, if the session uses our `DDURLSessionDelegate`, the swizzler will use it to talk with interceptor. If it doesn't, the swizzler will not do anything by design (check the `testGivenNonInterceptedSession_itDoesntCallInterceptor` test in `URLSessionSwizzlerTests.swift`).

> even if we let our users implement their own `InterceptorType`, we won't be calling their interceptors as their `session.delegate` is `Alamofire.SessionDelegate` type.

The idea in this PR is not to ask users to implement their interceptor type, but to expose the interceptor we already implement in the SDK. We expose this 3 methods in public API:
```swift
/// Notifies the `URLSessionTask` creation.
func taskCreated(task: URLSessionTask)

/// Notifies the `URLSessionTask` metrics collection.
func taskMetricsCollected(task: URLSessionTask, metrics: URLSessionTaskMetrics)

/// Notifies the `URLSessionTask` completion.
func taskCompleted(task: URLSessionTask, error: Error?)
```
so that the users can notify their `URLSessionTask` lifecycle. One more API is exposed to let them intercept the request for tracing headers injection:
```swift
/// Modifies the `URLRequest` before the `URLSessionTask` is created.
func modify(request: URLRequest) -> URLRequest
```

In [the Alamofire's "Advanced Usage" docs](https://github.com/Alamofire/Alamofire/blob/master/Documentation/AdvancedUsage.md) they clearly explain how custom network interception can be implemented using `EventMonitor` and `RequestInterceptor`. It fits perfectly the API of our interceptor, as I explained in the description of this PR.